### ISSUE TITLE
[BUG] Subdomains seem to be ignored for white listed hosts

### ISSUE THREAD

**Issue by boernsen-development** [IngoZenz/personaldnsfilter#254]
Description
---
Subdomains seem to be ignored for the white listed hosts in additionalHosts.txt even though the comment states it considers them. Maybe the comments in additionalHosts.txt are just not exact. Right now I understand that subdomains should be white listed as well and that more specific rules (such as a simple white list entry) have higher priority than blacklist rules. This seems to work for the domain itself, but not for subdomains.

Maybe I have overseen something?

Anyway, this would be quiet useful though when someone (like me) wants to block a range of hosts (or even everything) by default and then only allow specific hosts by white listing them.

I could imagine there is an easy fix.

Below is the reproduction recipe and log. I hope it helps. Feel free to ask me anything.

I tagged it as bug, hope it is correct and that I didn't oversee anything. Feel free to change it.

Test versions
---
personaldnsfilter 1.50.53.5 (in proxy mode)
Netguard 2.303 (providing the VPN and redirecting everything on port 53 to personaldnsfilter port 5300)

Config
---
> Filter lists: **None**
> Additional hosts:
```
# Add your own additional hosts here!
# Supports blacklist, whitelist and map to custom IP

# WARNING! You must not add huge lists here!
# This is only for overruling the configured filter.

# Format for blacklist: 1 host name per line (sample below).
# Wildcard Character "*" can be used for host blocking.
# Host name and all sub domains will be blocked,
# unless there is a more specific rule.

*google*

# In order to white list a specific host, use "!" prefix.
# Wildcard Character "*" can be used for whitelisting.
# Host name and all sub domains will be whitelisted,
# unless there is a more specific rule.

!googlevideo.com

# PRIORITIES in case of conflicting entries
# 1. Explicit entries without wildcards
# 2. Wildcard entries prioritized along the sequence in this file
# 3. Downloaded blocklist

# In order to forward to custom ip add a ">" in front of the host,
# followed by whitespace and mapped IP.

>router.home 192.168.2.1
```
> CNAME cloaking protection: **ON**
> DNS proxy mode without local VPN: **ON**
> Allow only local DNS proxy requests: **ON**
> Root mode without local VPN: **OFF**
> Prevent device sleep: **OFF**

Full config: [whitelist_bug.zip](https://github.com/IngoZenz/personaldnsfilter/files/10143978/whitelist_bug.zip)

Reproduction
---

- Apply config above
- Use any app (I used a browser) to open:
- - googlevideo.com
- - google.com
- - test.googlevideo.com
- As you can see in the log below:
- - googlevideo.com is correctly allowed :heavy_check_mark:
- - google.com is correctly blocked :heavy_check_mark:
- - test.googlevideo.com is blocked incorrectly :x:
- 
Log
---

> ***Initializing personalDNSfilter Version 1505305!***
> Using directory: /storage/emulated/0/Android/data/dnsfilter.android/files/PersonalDNSFilter/
> Acquired WIFI lock and partial wake lock!
> Released WIFI lock and partial wake lock!
> DNSFilterProxy running on port 5300!
> 20:51:26
> Selected DNS: (119ms) [5.9.164.112]::853::DOT
> DNS filter: Next filter reload:Thu Jan 01 01:00:00 GMT+01:00 1970
> DNS filter: Reloading hosts filter ...
> Updating filter completed!
> Reading filter file and building index...!
> Building index for 0 entries...!
> Persisting index for 0 entries...!
> Index contains 0 unique entries!
> Processing new filter file completed!
> Reloading hosts filter ... completed!
> DNS filter: Next filter reload:Fri Dec 09 20:51:26 GMT+01:00 2022
> 20:51:58
> ✓  googlevideo.com
> 20:52:00
> ✗  www.google.com
> 20:52:15
> ✗  test.googlevideo.com

**Comment by boernsen-development** [comment]
I just found a workaround:
Instead of putting `!googlevideo.com` below `*google*` put `!*googlevideo.com` above it. It has to be wildcarded and above.

**Comment by IngoZenz** [comment]
yes it seems it works as designed. In your first sample, there was no explicit entry for test.googlevideo.com, therefore it was blocked via the *google* rule hitting first. In your 2nd sample it hits the !*googlevideo.com rule first, thererfor it is whitelisted, whereas e.g. googleads.com would be blocked by *google* rule.

**Comment by boernsen-development** [comment]
So is `!googlevideo.com` not considered explicit or is it only considered explicit for googlevideo.com, but not for any sub-domain? I think this is what I misunderstood from:
```
# PRIORITIES in case of conflicting entries
# 1. Explicit entries without wildcards
# 2. Wildcard entries prioritized along the sequence in this file
# 3. Downloaded blocklist
```

**Comment by IngoZenz** [comment]
Exactly... Expicit means that there ia an explicit entry for a domain /sub domain

**Comment by boernsen-development** [comment]
Thanks for clarifying!

Technically, I would say this issue can be closed. Only a few suggestions for _additionalHosts.txt_ I would like to leave here for consideration to avoid confusion for others:

- mention the role of subdomains in the description of priorities as it seems that automatically derived subdomains are neither treated as explicit nor as wildcard
- move the whitelist section of the default file to the top and/or mention that if one wants to whitelist subdomains of matching wildcarded blacklist entries, this has to be done at the top AND using wildcards

**Comment by mosajjal** [comment]
another question that might be relevant here or needs its own issue: does "forward to custom ip" support wildcards? I'm keen to use this project as a companion piece to my SNI proxy and hard-coding some dns suffixes will save the users a bunch of useless road trips to the server

**Comment by IngoZenz** [comment]
Today the custom IP forwards do not support wildcards.
What is possible is to change the IP used for filtered hosts - maybe that's an option.

**Comment by mosajjal** [comment]
yes I saw a few "127.0.0.1" strings hardcoded in the app. would you be open to a PR for a setting to customize that? it'll solve my issue for sure :) could be a good one for some honeypot/fakenet implementations too

**Comment by IngoZenz** [comment]
You can customize it already today via following dnsfilter.conf settings:
```
# Returned IP for resolving blocked host
ipV4BlockedHost = 0.0.0.0
ipV6BlockedHost = ::
```

**Comment by mosajjal** [comment]
oh nice. yes just saw it and it works flawlessly :) thanks!

**Comment by zwrmueller** [comment]
> Today the custom IP forwards do not support wildcards

Hi,
is it possible to forward 1 host to 2 IPs?

Like:
> myhost 192.168.1.21
> myhost 192.168.1.22

That example above seems not to work, as (allways) the last entry is used ...

**Comment by IngoZenz** [comment]
it supports only 1 forward per host per IP4 / IP6

**Comment by zwrmueller** [comment]
Ok, thanx.
Would love it to see it the other way round, but thanx for quick response.

**Comment by IngoZenz** [comment]
There is an explicit entry for keep.google.com, therefore it is taken. Explicit entries have highest prio
